text/TextRenderer.cc

```diff
@@ -0,0 +1,203 @@
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "text/TextRenderer.h"
+
+#include "font.inc"
+
+void TextRenderer::setupTexture() {
+    glGenTextures(1, &fontTexture);
+
+    // Avoid interfering with caller's assumptions
+    glActiveTexture(GL_TEXTURE0);
+    GLint oldTexture;
+    glGetIntegerv(GL_TEXTURE_BINDING_2D, &oldTexture);
+
+    glBindTexture(GL_TEXTURE_2D, fontTexture);
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_RED, font.atlasWidth, font.atlasHeight,
+                 0, GL_RED, GL_UNSIGNED_BYTE, font.data.data());
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+
+    glBindTexture(GL_TEXTURE_2D, oldTexture);
+}
+
+void TextRenderer::setupBuffers() {
+    glGenVertexArrays(1, &vao);
+    glBindVertexArray(vao);
+    glGenBuffers(1, &vbo);
+    glBindBuffer(GL_ARRAY_BUFFER, vbo);
+    const GLfloat vertices[] = {
+        0, 0,
+        1, 0,
+        0, 1,
+        1, 1,
+    };
+    glBufferData(GL_ARRAY_BUFFER, sizeof vertices, vertices, GL_STATIC_DRAW);
+    constexpr GLuint attribIndex = 0;
+    constexpr int coordsPerVertex = 2;
+    glVertexAttribPointer(attribIndex, coordsPerVertex, GL_FLOAT, false, 0, 0);
+    glEnableVertexAttribArray(attribIndex);
+    glBindVertexArray(0);
+}
+
+void TextRenderer::setupProgram() {
+    const auto vertexShader = glCreateShader(GL_VERTEX_SHADER);
+    const char*const vertexShaderSrc = R"(#version 330
+uniform mat4 mvp;
+uniform vec2 texCoordIn;
+uniform vec2 charSizeInTexture;
+layout(location = 0) in vec4 vertex;
+out vec2 texCoord;
+void main()
+{
+    gl_Position = mvp*vertex;
+    texCoord = texCoordIn+charSizeInTexture*(vec2(vertex.x, 1-vertex.y));
+}
+)";
+    glShaderSource(vertexShader, 1, &vertexShaderSrc, nullptr);
+    glCompileShader(vertexShader);
+    const auto fragmentShader = glCreateShader(GL_FRAGMENT_SHADER);
+    const char*const fragmentShaderSrc = R"(#version 330
+uniform vec3 textColor;
+uniform sampler2D font;
+in vec2 texCoord;
+layout(location = 0) out vec4 color;
+void main()
+{
+    color = vec4(textColor.rgb, 1)*texture(font, texCoord).rrrr;
+}
+)";
+    glShaderSource(fragmentShader, 1, &fragmentShaderSrc, nullptr);
+    glCompileShader(fragmentShader);
+    program = glCreateProgram();
+    glAttachShader(program, vertexShader);
+    glAttachShader(program, fragmentShader);
+    glLinkProgram(program);
+    glDetachShader(program, fragmentShader);
+    glDeleteShader(fragmentShader);
+    glDetachShader(program, vertexShader);
+    glDeleteShader(vertexShader);
+}
+
+void TextRenderer::drawChar(char c, int x, int y,
+                            int viewportWidth, int viewportHeight) {
+    if (c<0x20 || c>0x7e) c = '?';
```

Comment:
Please use spaces around operators, throughout this file ("c < 0x20", "((c >> 4) - 2) / charH", etc)

Reply:
Any way to double-check this? CppLint (on Kubuntu 14.04) seems to be quite happy with current state...